R/DBConnection.R

```diff
@@ -106,6 +106,97 @@ setGeneric("dbSendQuery",
   valueClass = "DBIResult"
 )
 
+#' Sentinel value to indicate the early termination of
+#' \code{\link{dbGetChunkedQuery}}
+#' @export
+dbBreak <- structure(list(), class = "dbi_break_sentinel")
+
+dbFetchChunked <- function(rs, n, callback) {
+  while (TRUE) {
+    chunk <- dbFetch(rs, n = n)
+    if (nrow(chunk) == 0) {
+      return(chunk)
+    }
+    result <- callback(chunk)
```

Comment:
I'm adding an extra argument to the callback to allow for `callback(chunk, index = rowsSoFar)`. But this got me thinking, are we okay with the fact that the callback cannot take any extra arguments at all (no `...`)? The way things are currently implemented, this is not possible because the `...` are taken up by the `dbSendQuery` method:

```r
#' @export
setMethod("dbGetChunkedQuery", signature("DBIConnection", "character"),
  function(conn, statement, callback, n, ...) {
    rs <- dbSendQuery(conn, statement, ...)
    on.exit(dbClearResult(rs))
    dbFetchChunked(rs, callback, n)
  }
)
```